src/modules/core/filter_audiochannels.c

```diff
@@ -23,216 +23,416 @@
+				memcpy( p, in_buffer, size_avail );
+			}
 		}
 	}
-	else if ( channels_avail > *channels )
+	else if ( in_channels > out_channels )
 	{
-		int size = mlt_audio_format_size( *format, *samples, *channels );
-		int16_t *new_buffer = mlt_pool_alloc( size );
 		int i, j;
 
-		// Drop all but the first *channels
-		if ( *format == mlt_audio_s16 )
+		// Drop all but the first out_channels
+		if ( format == mlt_audio_s16 )
 		{
-			for ( i = 0; i < *samples; i++ )
-				for ( j = 0; j < *channels; j++ )
-					new_buffer[ ( i * *channels ) + j ]	= ((int16_t*)(*buffer))[ ( i * channels_avail ) + j ];
+			int16_t *p = (int16_t*) out_buffer;
+			for ( i = 0; i < samples; i++ )
+				for ( j = 0; j < out_channels; j++ )
+					p[ ( i * out_channels ) + j ]	= ((int16_t*)(in_buffer))[ ( i * in_channels ) + j ];
 		}
-		else if ( *format == mlt_audio_s32le || *format == mlt_audio_f32le )
+		else if ( format == mlt_audio_s32le || format == mlt_audio_f32le )
 		{
-			int32_t *p = (int32_t*) new_buffer;
-			for ( i = 0; i < *samples; i++ )
-				for ( j = 0; j < *channels; j++ )
-					p[ ( i * *channels ) + j ]	= ((int32_t*)(*buffer))[ ( i * channels_avail ) + j ];
+			int32_t *p = (int32_t*) out_buffer;
+			for ( i = 0; i < samples; i++ )
+				for ( j = 0; j < out_channels; j++ )
+					p[ ( i * out_channels ) + j ]	= ((int32_t*)(in_buffer))[ ( i * in_channels ) + j ];
 		}
-		else if ( *format == mlt_audio_u8 )
+		else if ( format == mlt_audio_u8 )
 		{
-			uint8_t *p = (uint8_t*) new_buffer;
-			for ( i = 0; i < *samples; i++ )
-				for ( j = 0; j < *channels; j++ )
-					p[ ( i * *channels ) + j ]	= ((uint8_t*)(*buffer))[ ( i * channels_avail ) + j ];
+			uint8_t *p = (uint8_t*) out_buffer;
+			for ( i = 0; i < samples; i++ )
+				for ( j = 0; j < out_channels; j++ )
+					p[ ( i * out_channels ) + j ]	= ((uint8_t*)(in_buffer))[ ( i * in_channels ) + j ];
 		}
 		else
 		{
 			// non-interleaved - s32 or float
-			memcpy( new_buffer, *buffer, size );
+			memcpy( out_buffer, in_buffer, mlt_audio_format_size( format, samples, out_channels ) );
 		}
-		// Update the audio buffer now - destroys the old
-		mlt_frame_set_audio( frame, new_buffer, *format, size, mlt_pool_release );
-		*buffer = new_buffer;
 	}
+}
+
+static int filter_get_audio( mlt_frame frame, void **buffer, mlt_audio_format *format, int *frequency, int *channels, int *samples )
+{
+	// Used to return number of channels in the source
+	int channels_avail = *channels;
+
+	// Get the producer's audio
+	int error = mlt_frame_get_audio( frame, buffer, format, frequency, &channels_avail, samples );
+	if ( error ) return error;
+
+	if ( channels_avail == *channels )
+	{
+		return error;
+	}
+
+	int size = mlt_audio_format_size( *format, *samples, *channels );
+	float *new_buffer = mlt_pool_alloc( size );
+	char *convert = mlt_properties_get( MLT_FRAME_PROPERTIES( frame ), "audiochannel.convert" );
+	if ( channels_avail > 8 || ( convert && !strcmp( convert, "dupdrop" ) ) )
```

Comment:
Default mode is "mix". The new surround mixing will be used unless the user/application specifies "channelconvert=dupdrop" on the consumer. I chose mix as the default because that makes avcodecs that don't support codec downmix (like AAC) behave the same as avcodecs that do support codec downmix (like ac3).

Reply:
If we publish the frame property as "consumer_channels_convert", then the avformat producer could read that and if "dupdrop" then override the request_channel_layout to prevent the need to supply that property explicitly on the producer to get desired result. Alternatively, we can always set request_channel_layout from av_get_default_channel_layout(codec_context->channels) to defer all audio channel conversion to our filter.